README.md
# keycloak-http-auth
Validate auth and enforce ACLs for HTTP-based Keycloak token requests. Ideal for answering nginx auth requests.

Because this is meant for non-interactive use, failures only return 403.

On success, these http headers are returned:
* REMOTE_USER: the `posix.username` or `upn` or `preferred_username` in the token (required)
* X_UID: the `posix.uid` in the token (optional)
* X_GID: the `posix.gid` in the token (optional)

UID and GID are useful for POSIX filesystem access, for example [https://unix.stackexchange.com/a/489744](https://unix.stackexchange.com/a/489744).

## Configuration

Primary configuration is via environment variables:

* ISSUERS: the issuers, comma-separated
* AUDIENCE: the audience
* BASE_PATH: the base path of the server, before any authorized paths in the token


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.0.16 (2023-04-18)


## v1.0.15 (2022-08-26)


## v1.0.14 (2022-08-24)


## v1.0.13 (2022-08-18)


## v1.0.12 (2022-01-25)


## v1.0.11 (2022-01-25)


## v1.0.10 (2022-01-24)


## v1.0.9 (2022-01-24)


## v1.0.8 (2022-01-21)


## v1.0.7 (2022-01-20)


## v1.0.6 (2022-01-20)


## v1.0.5 (2022-01-20)


## v1.0.4 (2022-01-13)



tests/util.py
import json
import jwt
import pytest
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.rsa import generate_private_key

from rest_tools.utils import Auth


@pytest.fixture(scope="session")
def gen_keys():
    priv = generate_private_key(65537, 2048)
    pub = priv.public_key()
    return (priv, pub)

@pytest.fixture(scope="session")
def gen_keys_bytes(gen_keys):
    priv, pub = gen_keys

    priv_pem = priv.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption()
    )
    pub_pem = pub.public_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PublicFormat.SubjectPublicKeyInfo
    )
    print(priv_pem, pub_pem)
    return (priv_pem, pub_pem)

@pytest.fixture(scope="session")
def gen_jwk(gen_keys):
    jwk = json.loads(jwt.algorithms.RSAAlgorithm.to_jwk(gen_keys[1]))
    jwk['kid'] = 'testing'
    return jwk

@pytest.fixture
def make_token(gen_keys_bytes):
    def func(posix, issuer, audience):
        auth = Auth(gen_keys_bytes[0], pub_secret=gen_keys_bytes[1], algorithm='RS256', issuer=issuer)
        return auth.create_token('testing', payload={
            'scope': 'posix',
            'aud': audience,
            'posix': posix,
        }, headers={'kid': 'testing'})
    yield func


tests/test_server.py
import socket
import asyncio

import pytest
from requests.exceptions import HTTPError
from rest_tools.client import AsyncSession
import requests_mock
import pytest_asyncio

from keycloak_http_auth.server import create_server

from .util import *

@pytest.fixture
def port():
    """Get an ephemeral port number."""
    # https://unix.stackexchange.com/a/132524
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.bind(('', 0))
    addr = s.getsockname()
    ephemeral_port = addr[1]
    s.close()
    return ephemeral_port

@pytest_asyncio.fixture
async def server(monkeypatch, port, make_token, gen_jwk, tmp_path):
    monkeypatch.setenv('DEBUG', 'True')
    monkeypatch.setenv('PORT', str(port))

    monkeypatch.setenv('ISSUERS', 'issuer')
    monkeypatch.setenv('AUDIENCE', 'aud')
    base_path = tmp_path / 'base'
    base_path.mkdir()
    monkeypatch.setenv('BASE_PATH', str(base_path))
    monkeypatch.setenv('KEYCLOAK_URL', 'http://foo')
    monkeypatch.setenv('KEYCLOAK_REALM', 'testing')

    with requests_mock.Mocker(real_http=True) as m:
        m.get('http://foo/auth/realms/testing/.well-known/openid-configuration', text=json.dumps({
            'token_endpoint': 'http://foo/auth/realms/testing/token',
            'jwks_uri': 'http://foo/auth/realms/testing/certs',
        }))
        m.get('http://foo/auth/realms/testing/certs', text=json.dumps({
            'keys': [gen_jwk],
        }))

        s = create_server()

    def fn(posix):
        token = make_token(posix, 'issuer', 'aud')
        session = AsyncSession(retries=0)
        async def fn2(method, path, headers={}):
            url = f'http://localhost:{port}{path}'
            headers['Authorization'] = 'Bearer '+token
            ret = await asyncio.wrap_future(session.request(method, url, timeout=0.1, headers=headers))
            ret.raise_for_status()
            return ret
        return fn2, base_path

    try:
        yield fn
    finally:
        await s.stop()

@pytest.mark.asyncio
async def test_server_health(server):
    client, _ = server({'username': 'foo', 'uid': 1000, 'gid': 1001})
    ret = await client('GET', '/healthz')

@pytest.mark.asyncio
async def test_server_root_path(server):
    client, base_path = server({'username': 'foo', 'uid': 1000, 'gid': 1001})
    with open(base_path / 'foo', 'w') as f:
        f.write('foo')
    ret = await client('GET', '/', headers={
        'X-Original-Method': 'GET',
        'X-Original-URI': '/foo',
    })
    assert ret.headers['REMOTE_USER'] == 'foo'
    assert ret.headers['X_UID'] == '1000'
    assert ret.headers['X_GID'] == '1001'
    assert ret.headers['X_GROUPS'] == '1001'

@pytest.mark.asyncio
async def test_server_post(server):
    client, _ = server({'username': 'foo', 'uid': 1000, 'gid': 1000})
    with pytest.raises(HTTPError, match='405'):
        await client('POST', '/', headers={
            'X-Original-Method': 'GET',
            'X-Original-URI': '/base/',
        })

@pytest.mark.asyncio
async def test_server_missing_user(server):
    client, _ = server({})
    with pytest.raises(HTTPError, match='400'):
        await client('GET', '/', headers={
            'X-Original-Method': 'GET',
            'X-Original-URI': '/base/',
        })


integration_tests/test_nginx.py
import http.server
import logging
import os
from pathlib import Path
import shutil
import socket
import stat
import subprocess
from threading import Thread
import time

import pytest
import requests

from .util import *


def port():
    """Get an ephemeral port number."""
    # https://unix.stackexchange.com/a/132524
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.bind(('', 0))
    addr = s.getsockname()
    ephemeral_port = addr[1]
    s.close()
    return ephemeral_port

config = """
listen 0.0.0.0:{nginx_port};
location /auth {{
  internal;
  proxy_pass              http://127.0.0.1:{app_port}/;
  proxy_connect_timeout   {timeout}s;
  proxy_read_timeout      {timeout}s;
  proxy_send_timeout      {timeout}s;
  proxy_set_header        Content-Length "";
  proxy_set_header        X-Original-URI $request_uri;
  proxy_set_header        X-Original-Method $request_method;
}}
"""

config_health = """
server {{
  listen 0.0.0.0:{health_port};
  location = /basic_status {{
    stub_status;
  }}
}}
"""

@pytest.fixture(scope="session")
def fake_server():
    class Handler(http.server.BaseHTTPRequestHandler):
        def do_GET(self):
            self.wfile.write(b'HTTP/1.0 200 OK\r\n')
            self.send_header('REMOTE_USER', 'user')
            self.send_header('X_UID', '123')
            self.send_header('X_GID', '456')
            self.send_header('X_GROUPS', '456,789')
            self.end_headers()
            self.wfile.write(b'\r\n')
    app_port = port()
    s = http.server.HTTPServer(('', app_port), Handler)
    t = Thread(target=s.serve_forever, daemon=True)
    t.start()
    yield app_port

@pytest.fixture(scope="session")
def nginx(tmp_path_factory, fake_server):
    workdir = Path(__file__).parent.parent
    subprocess.run(['docker', 'build', '-t', 'wipac/keycloak-http-auth:testing',
                    '-f', 'Dockerfile_nginx', f'{workdir}'], check=True, cwd=workdir)

    test_volume = tmp_path_factory.mktemp('cache')
    nginx_config = tmp_path_factory.mktemp('conf') / 'config.conf'
    nginx_health_config = tmp_path_factory.mktemp('conf') / 'health.conf'
    nginx_port = port()
    health_port = port()
    app_port = fake_server

    nginx_config.write_text(config.format(nginx_port=nginx_port, app_port=app_port, timeout=1))
    nginx_health_config.write_text(config_health.format(health_port=health_port))

    with subprocess.Popen(['docker', 'run', '--rm', '--network=host', '--name', 'test_nginx_integration',
                           '-v', f'{nginx_config}:/etc/nginx/custom/webdav.conf:ro',
                           '-v', f'{nginx_health_config}:/etc/nginx/sites-enabled/health.conf:ro',
                           '-v', f'{test_volume}:/mnt/data:rw', 'wipac/keycloak-http-auth:testing']) as p:
        # wait for server to come up
        for i in range(10):
            try:
                socket.create_connection(('localhost', nginx_port), 0.1)
            except ConnectionRefusedError:
                if i >= 9:
                    raise
                else:
                    time.sleep(.1)

        def fn(*args):
            return subprocess.run(['docker', 'exec', 'test_nginx_integration']+list(args), check=True, cwd=workdir, capture_output=True).stdout

        try:
            yield {
                'data': test_volume,
                'nginx_port': nginx_port,
                'health_port': health_port,
                'app_port': app_port,
                'exec_in_container': fn,
            }
        finally:
            p.terminate()


@pytest.fixture(autouse=True, scope='function')
def clear_test_volume(nginx):
    nginx["data"].chmod(0o777)
    try:
        uid, gid = (int(x) for x in nginx['exec_in_container']('stat','-c','%u %g','/mnt/data').split())
        logging.info(f'uid:gid starts as {uid}:{gid}')
        yield
    finally:
        logging.info('cleaning dir')
        nginx['exec_in_container']('chown', '-R', f'{uid}:{gid}', '/mnt/data')
        for child in nginx['data'].iterdir():
            if child.is_dir():
                shutil.rmtree(child)
            else:
                child.unlink()
        logging.debug('ls: %r', nginx['exec_in_container']('ls','-al','/mnt/data'))

def test_health(nginx):
    r = requests.get(f'http://localhost:{nginx["health_port"]}/basic_status')
    assert r.status_code == 200
    assert 'Active connections' in r.text

def test_index(nginx):
    r = requests.get(f'http://localhost:{nginx["nginx_port"]}/')
    assert r.status_code == 200

def test_missing(nginx):
    r = requests.get(f'http://localhost:{nginx["nginx_port"]}/missing')
    assert r.status_code == 404

def test_read(nginx):
    data = b'foo bar baz'
    nginx["data"].chmod(0o777)
    (nginx["data"] / "test").write_bytes(data)

    r = requests.get(f'http://localhost:{nginx["nginx_port"]}/data/test')
    r.raise_for_status()
    assert r.content == data

def test_read_bad_perms(nginx):
    data = b'foo bar baz'
    nginx["data"].chmod(0o777)
    (nginx["data"] / "test").write_bytes(data)
    (nginx["data"] / "test").chmod(0o600)

    r = requests.get(f'http://localhost:{nginx["nginx_port"]}/data/test')
    with pytest.raises(Exception):
        r.raise_for_status()

def test_read_group(nginx):
    data = b'foo bar baz'
    nginx["data"].chmod(0o777)
    (nginx["data"] / "test").write_bytes(data)
    (nginx["data"] / "test").chmod(0o660)
    nginx['exec_in_container']('chgrp','789','/mnt/data/test')

    r = requests.get(f'http://localhost:{nginx["nginx_port"]}/data/test')
    r.raise_for_status()
    assert r.content == data

def test_read_bad_group(nginx):
    data = b'foo bar baz'
    nginx["data"].chmod(0o777)
    (nginx["data"] / "test").write_bytes(data)
    (nginx["data"] / "test").chmod(0o660)
    nginx['exec_in_container']('chgrp','12345','/mnt/data/test')

    r = requests.get(f'http://localhost:{nginx["nginx_port"]}/data/test')
    with pytest.raises(Exception):
        r.raise_for_status()

def test_write(nginx):
    data = b'foo bar baz'
    nginx["data"].chmod(0o777)

    r = requests.put(f'http://localhost:{nginx["nginx_port"]}/data/test', data=data)
    r.raise_for_status()

    out = nginx['exec_in_container']('cat', '/mnt/data/test')
    assert out == data
    out = nginx['exec_in_container']('stat','-c','%u %g','/mnt/data/test')
    assert out.strip() == b'123 456'

def test_write_group(nginx):
    data = b'foo bar baz'
    nginx["data"].chmod(0o770)
    nginx['exec_in_container']('chgrp','789','/mnt/data')
    out = nginx['exec_in_container']('stat','-c','%A','/mnt/data')
    assert out.strip() == b'drwxrwx---'

    r = requests.put(f'http://localhost:{nginx["nginx_port"]}/data/test', data=data)
    r.raise_for_status()

    out = nginx['exec_in_container']('cat', '/mnt/data/test')
    assert out == data
    out = nginx['exec_in_container']('stat','-c','%A %u %g','/mnt/data/test')
    assert out.strip() == b'-rw-rw---- 123 456'

def test_write_group_sticky(nginx):
    data = b'foo bar baz'
    nginx["data"].chmod(0o770)
    nginx['exec_in_container']('chgrp','789','/mnt/data')
    nginx['exec_in_container']('chmod','g+s','/mnt/data')
    out = nginx['exec_in_container']('stat','-c','%A','/mnt/data')
    assert out.strip() == b'drwxrws---'

    r = requests.put(f'http://localhost:{nginx["nginx_port"]}/data/test', data=data)
    r.raise_for_status()

    out = nginx['exec_in_container']('cat', '/mnt/data/test')
    assert out == data
    out = nginx['exec_in_container']('stat','-c','%A %u %g','/mnt/data/test')
    assert out.strip() == b'-rw-rw---- 123 789'

def test_write_subdir(nginx):
    data = b'foo bar baz'
    nginx["data"].chmod(0o770)
    nginx['exec_in_container']('chgrp','789','/mnt/data')
    nginx['exec_in_container']('chmod','g+s','/mnt/data')
    out = nginx['exec_in_container']('stat','-c','%A','/mnt/data')
    assert out.strip() == b'drwxrws---'

    r = requests.put(f'http://localhost:{nginx["nginx_port"]}/data/sub1/sub2/test', data=data)
    r.raise_for_status()

    out = nginx['exec_in_container']('cat', '/mnt/data/sub1/sub2/test')
    assert out == data
    out = nginx['exec_in_container']('stat','-c','%A %u %g','/mnt/data/sub1/sub2/test')
    assert out.strip() == b'-rw-rw---- 123 789'

def test_write_bad_group(nginx):
    data = b'foo bar baz'
    nginx["data"].chmod(0o770)
    nginx['exec_in_container']('chown','12345:12345','/mnt/data')
    out = nginx['exec_in_container']('stat','-c','%A','/mnt/data')
    assert out.strip() == b'drwxrwx---'

    r = requests.put(f'http://localhost:{nginx["nginx_port"]}/data/test', data=data)
    with pytest.raises(Exception):
        r.raise_for_status()

    with pytest.raises(Exception):
        out = nginx['exec_in_container']('stat','/mnt/data/test')
        logging.warning('stat output: %r', out)
    #f = (nginx["data"] / "test")
    #assert not f.exists()

def test_write_bad_perms(nginx):
    data = b'foo bar baz'
    nginx["data"].chmod(0o555)

    r = requests.put(f'http://localhost:{nginx["nginx_port"]}/data/test', data=data)
    with pytest.raises(Exception):
        r.raise_for_status()

    f = (nginx["data"] / "test")
    assert not f.exists()

def test_delete(nginx):
    data = b'foo bar baz'
    nginx["data"].chmod(0o777)

    r = requests.put(f'http://localhost:{nginx["nginx_port"]}/data/test', data=data)
    r.raise_for_status()

    r = requests.delete(f'http://localhost:{nginx["nginx_port"]}/data/test')
    r.raise_for_status()

    f = (nginx["data"] / "test")
    assert not f.exists()

def test_mkdir(nginx):
    data = b'foo bar baz'
    nginx["data"].chmod(0o777)

    r = requests.request('MKCOL', f'http://localhost:{nginx["nginx_port"]}/data/test/')
    r.raise_for_status()

    f = (nginx["data"] / "test")
    assert f.is_dir()
    out = nginx['exec_in_container']('stat','-c','%u %g','/mnt/data/test')
    assert out.strip() == b'123 456'

def test_bad_methods(nginx):
    data = b'foo bar baz'
    for m in ('POST', 'PATCH'):
        r = requests.request(m, f'http://localhost:{nginx["nginx_port"]}/data/test/', data=data)
        assert r.status_code == 405


keycloak_http_auth/__main__.py
import asyncio
import logging

from rest_tools.utils import from_environment

from .server import create_server

# handle logging
setlevel = {
    'CRITICAL': logging.CRITICAL,  # execution cannot continue
    'FATAL': logging.CRITICAL,
    'ERROR': logging.ERROR,  # something is wrong, but try to continue
    'WARNING': logging.WARNING,  # non-ideal behavior, important event
    'WARN': logging.WARNING,
    'INFO': logging.INFO,  # initial debug information
    'DEBUG': logging.DEBUG  # the things no one wants to see
}

default_config = {
    'LOG_LEVEL': 'INFO',
}
config = from_environment(default_config)
if config['LOG_LEVEL'].upper() not in setlevel:
    raise Exception('LOG_LEVEL is not a proper log level')
logformat = '%(asctime)s %(levelname)s %(name)s %(module)s:%(lineno)s - %(message)s'

logging.basicConfig(format=logformat, level=setlevel[config['LOG_LEVEL'].upper()])

# start server
create_server()
asyncio.get_event_loop().run_forever()


keycloak_http_auth/server.py
"""
Server for keycloak token http auth
"""

import logging

from tornado.web import HTTPError
from rest_tools.server import RestServer, RestHandler, RestHandlerSetup, authenticated, catch_error
from rest_tools.utils import from_environment


class Main(RestHandler):
    @authenticated
    @catch_error
    async def get(self, *args):
        method = self.request.headers.get('X-Original-Method', '')
        path = self.request.headers.get('X-Original-URI', '')
        token = self.auth_data

        # test basic token properties
        if 'sub' not in token:
            raise HTTPError(403, reason='sub not in token')

        username = token.get('posix', {}).get('username', None)
        if not username:
            username = token.get('upn', None)
        if not username:
            username = token.get('preferred_username', None)
        if not username:
            raise HTTPError(400, 'username missing from token')

        logging.info(f'request for user {username}: {method}:{path}')

        # check for posix info
        uid = token.get('posix', {}).get('uid', None)
        if not uid:
            raise HTTPError(403, reason='posix.uid missing from token')
        gid = token.get('posix', {}).get('gid', None)
        if not gid:
            raise HTTPError(403, reason='posix.gid missing from token')
        gids = set(token.get('posix', {}).get('group_gids', []))
        if gid:
            gids.add(gid)

        # if you want to do other checks, add them here

        # set nginx uid and gid
        self.set_header('REMOTE_USER', username)
        self.set_header('X_UID', uid)
        self.set_header('X_GID', gid)
        self.set_header('X_GROUPS', ','.join(f'{g}' for g in gids))
        self.write('')


class Health(RestHandler):
    def get(self):
        self.write('')


def create_server():
    default_config = {
        'HOST': 'localhost',
        'PORT': 8080,
        'DEBUG': False,
        'ISSUERS': None,
        'AUDIENCE': None,
        'KEYCLOAK_URL': None,
        'KEYCLOAK_REALM': 'IceCube',
    }
    config = from_environment(default_config)

    rest_config = {
        'debug': config['DEBUG'],
        'auth': {
            'openid_url': f'{config["KEYCLOAK_URL"]}/auth/realms/{config["KEYCLOAK_REALM"]}',
            'audience': config['AUDIENCE'].split(','),
            'issuers': config['ISSUERS'].split(','),
        }
    }
    kwargs = RestHandlerSetup(rest_config)

    server = RestServer(debug=config['DEBUG'])
    server.add_route('/healthz', Health, kwargs)
    server.add_route(r'/(.*)', Main, kwargs)

    server.startup(address=config['HOST'], port=config['PORT'])

    return server
